libraries/node-core-library/src/PackageJsonLookup.ts

```diff
@@ -40,56 +57,152 @@ export class PackageJsonLookup {
+  public tryGetPackageJsonFilePathFor(fileOrFolderPath: string): string | undefined {
+    const packageJsonFolder: string | undefined = this.tryGetPackageFolderFor(fileOrFolderPath);
+    if (!packageJsonFolder) {
+      return undefined;
     }
+    return path.join(packageJsonFolder, FileConstants.PackageJson);
+  }
 
-    const packageJson: IPackageJson = JsonFile.load(path.join(packageJsonPath, 'package.json')) as IPackageJson;
-    result = packageJson.name;
+  /**
+   * If the specified file or folder is part of a package, this loads and returns the
+   * associated package.json file.
+   *
+   * @remarks
+   * The package folder is determined using the same algorithm
+   * as {@link PackageJsonLookup.tryGetPackageFolderFor}.
+   *
+   * @param fileOrFolderPath - a relative or absolute path to a source file or folder
+   * that may be part of a package
+   * @returns an IPackageJson object, or undefined if the fileOrFolderPath does not
+   * belong to a package
+   */
+  public tryLoadPackageJsonFor(fileOrFolderPath: string): IPackageJson | undefined {
+    const packageJsonFilePath: string | undefined = this.tryGetPackageJsonFilePathFor(fileOrFolderPath);
+    if (!packageJsonFilePath) {
+      return undefined;
+    }
+    return this.loadPackageJson(packageJsonFilePath);
+  }
+
+  /**
+   * Loads the specified package.json file, if it is not already present in the cache.
+   *
+   * @remarks
+   * Unless {@link IPackageJsonLookupParameters.loadExtraFields} was specified,
+   * the returned IPackageJson object will contain a subset of essential fields.
+   * The returned object should be considered to be immutable; the caller must never
+   * modify it.
+   *
+   * @param jsonFilename - a relative or absolute path to a package.json file
+   */
+  public loadPackageJson(jsonFilename: string): IPackageJson {
+    if (!fsx.existsSync(jsonFilename)) {
+      throw new Error(`Input file not found: ${jsonFilename}`);
+    }
+
+    // Since this will be a cache key, follow any symlinks and get an absolute path
+    // to minimize duplication.  (Note that duplication can still occur due to e.g. character case.)
+    const normalizedFilePath: string = fsx.realpathSync(jsonFilename);
+
+    let packageJson: IPackageJson | undefined = this._packageJsonCache.get(normalizedFilePath);
+
+    if (!packageJson) {
+      const loadedPackageJson: IPackageJson = JsonFile.load(normalizedFilePath) as IPackageJson;
+
+      // Make sure this is really a package.json file.  CommonJS has fairly strict requirements,
+      // but NPM only requires "name" and "version"
+      if (!loadedPackageJson.name) {
+        throw new Error(`Error reading "${jsonFilename}":\n  `
+          + 'The required field "name" was not found');
+      }
+      if (!loadedPackageJson.version) {
+        throw new Error(`Error reading "${jsonFilename}":\n  `
+          + 'The required field "version" was not found');
+      }
+
+      if (this._loadExtraFields) {
```

Comment:
>_loadExtraFields [](start = 15, length = 16)

What about an optional parameter that lets you list out properties that you want?

Reply:
It's a good idea, but I'd like to wait until someone really needs it before designing it.

Reply:
It's a good idea, but I'd like to wait until someone really needs it before designing it.

---
In reply to: [175201618](https://github.com/Microsoft/web-build-tools/pull/564#discussion_r175201618) [](ancestors = 175201618)